v22.1/crdb-internal.md

```diff
@@ -578,6 +579,102 @@ WHERE metadata @> '{"db":"movr"}' AND (metadata @> '{"stmtTyp":"TypeDDL"}' OR me
 
 ~~~
 
+#### Detect suboptimal and regressed plans <span class="version-tag">New in v22.1</span>
+
+To detect suboptimal and regressed plans over time you can compare plans for the same query by extracting them from `statistics->'statistics'->'planGists'`.
+
+Suppose you wanted to compare plans of the following query:
+
+{% include copy-clipboard.html %}
+~~~ sql
+SELECT
+  name, count(rides.id) AS sum
+FROM
+  users JOIN rides ON users.id = rides.rider_id
+WHERE
+  rides.start_time BETWEEN '2018-12-31 00:00:00' AND '2020-01-01 00:00:00'
+GROUP BY
+  name
+ORDER BY
+  sum DESC
+LIMIT
+  10;
+~~~
+
+To decode plan gists stored in `statistics->'statistics'->'planGists'`, you use the `crdb_internal.decode_plan_gist` function. Because a plan is several lines, you should select only the `planGists` column and apply the function as follows; if you were to select more than one column the select would display only the first line of the plan.
+
+{% include copy-clipboard.html %}
+~~~ sql
+SELECT crdb_internal.decode_plan_gist(statistics->'statistics'->'planGists'->>0) AS plan FROM movr.crdb_internal.statement_statistics WHERE substring(metadata ->> 'query',1,35)='SELECT name, count(rides.id) AS sum';
+~~~
+
+~~~
+                    plan
+---------------------------------------------
+  • top-k
+  │ order
+  │
+  └── • group (hash)
+      │ group by: id
+      │
+      └── • hash join
+          │ equality: (id) = (rider_id)
+          │
+          ├── • scan
+          │     table: users@users_pkey
+          │     spans: FULL SCAN
+          │
+          └── • filter
+              │
+              └── • scan
+                    table: rides@rides_pkey
+                    spans: FULL SCAN
+
+~~~
+
+After you add an [index on the `start_time` column in the `rides` table](apply-statement-performance-rules.html#rule-2-use-the-right-index), you can see the change in the plans for the same query.
+
+{% include copy-clipboard.html %}
+~~~ sql
+SELECT crdb_internal.decode_plan_gist(statistics->'statistics'->'planGists'->>0) AS plan FROM movr.crdb_internal.statement_statistics WHERE substring(metadata ->> 'query',1,35)='SELECT name, count(rides.id) AS sum';
```

Comment:
Related to my previous question, I wonder if we can structure the output a bit better here. Going through this example, we just show two plans as the result with no other context:

<img width="1081" alt="image" src="https://user-images.githubusercontent.com/81193618/163034093-9901054f-c7b9-4a27-aa99-14dbd035a7b8.png">

It would be nice to group a plan as a single row in the result with additional context like the statement text and aggregation interval (when it was captured) as well.

Reply:
@stbof, i wrote a query using string_agg function that should clean up the output a bit. Let me know what you think. 

```
SELECT 
  statistics->'statistics'->'planGists'->>0 as plan_id,
  string_agg(  crdb_internal.decode_plan_gist(statistics->'statistics'->'planGists'->>0), '
  ')
AS plan,
substring(metadata ->> 'query',1,60) AS statement_text
FROM movr.crdb_internal.statement_statistics
WHERE substring(metadata ->> 'query',1,35)='SELECT name, count(rides.id) AS sum'
group by metadata ->> 'query', statistics->'statistics'->'planGists'->>0;
```

Reply:
OK. I see that that gives the side-by-side view you were looking for. I clearly need to beef up my SQL skills to be able to get output like this. I have a few questions:

- Do we need the plan ID in the output
- Do you want the statement text on the left or the right? 
- Do you want the aggregation interval as you mentioned in your first comment?
- Do we want to reverse order, so that we can see the plan _after_ the index is added?
- Do we want to show other statistics, so that the plan behavior is clear? If so, which one?

Reply:
I am no expert either - had to looke up how to merge multiple rows into a single row and column. There could be a better way to do this. Perhaps @cucaroach may know :)

- It's not needed but i couldn't get the query to work since we can't use the gist function in the group by. And since the plan_id represents the plan itself, it was the best candidate to ensure we had one row per plan.
- Yes, first column would be good!
- Yes, let's add the aggregation interval (probably take the max for the most recent time interval)
- Yes, that would be good. Perhaps order by latency desc so that the FULL SCAN is the first row
- Yes, that's a good idea. rowsRead and runLat are good stats to show the change after the index is added.


```
SELECT 
substring(metadata ->> 'query',1,60) AS statement_text,
  statistics->'statistics'->'planGists'->>0 as plan_id,
  string_agg(  crdb_internal.decode_plan_gist(statistics->'statistics'->'planGists'->>0), '
  ')
AS plan,
max(aggregated_ts) as timestamp_interval,
max(statistics -> 'statistics' -> 'rowsRead' -> 'mean') AS num_rows_read_mean,
max(statistics -> 'statistics' -> 'runLat' -> 'mean') AS runtime_latency_mean
FROM movr.crdb_internal.statement_statistics
WHERE substring(metadata ->> 'query',1,35)='SELECT name, count(rides.id) AS sum'
group by metadata ->> 'query', statistics->'statistics'->'planGists'->>0;
```

Reply:
No sorry, nothing comes to mind.   The side by side view is slick though!